packages/dds/tree/docs/main/merge-semantics.md

```diff
@@ -0,0 +1,354 @@
+-   The need to select a data model (i.e., how a document is structured) for an application so that the application's invariants are upheld by `SharedTree`'s merge semantics.
+
+For example, consider an application whose data model includes two arrays,
+with the invariant that the length of one array is expected to always be the same as the length of the other array.
+The application's editing code may attempt to keep the two arrays' length in sync by always adding to and removing from both array in equal measure.
+Despite that, `SharedTree`'s merge semantics are such that a scenario involving concurrent edits may still lead to a state where the arrays end up with different length.
+Understanding `SharedTree`'s merge semantics can help the application author anticipate this invariant violation or (failing that) diagnose it after the fact.
+Understanding `SharedTree`'s merge semantics will also enable the application author to understand how to remedy this danger,
+either by adopting a data model that prevents the issue (e.g., using a single array of pairs),
+or by changing the application's editing code (using transactions and constraints) to circumvent it.
+
+## How We Describe Merge Semantics
+
+The merge semantics of a given kind of edit (e.g., moving a node) define what the impact of the edit will be
+depending on what concurrent edits were sequenced before it.
+This could be captured in a set of rules of the form:
+`if <specific-concurrent-edit> was sequenced before this edit then the outcome of this edit will be <specific-outcome>`.
+Capturing the semantics of edits in this way would however be cumbersome because the set of possible concurrent edits and the set possible matching outcome if often open-ended.
+We instead describe the merge semantics of each kind of edit by describing its preconditions and postconditions.
+
+It is these preconditions and postconditions that define whether an edit's merge semantics make it suitable for a given scenario.
+
+### Preconditions
+
+The preconditions characterize what must be true for the edit to be considered valid.
+If those preconditions are not met, then the edit, and the whole transaction it is part of,
+is dropped: they have no effect.
+
+For example, imagine Alice inserts a node of type `Foo` in an array,
+with the precondition that by the time the edit is finally applied,
+this array allow must instances of type `Foo`.
+Concurrently to that, Bob edits the document schema such that the array in question no longer allows instances of type `Foo`,
+with the precondition that by the time the edit is finally applied,
+no instances of type `Foo` must exist in the array.
+If Alice's edit is sequenced first, then the node will successfully be inserted, and Bob's edit will be dropped.
+If Bob's edit is sequenced first, then the schema will successfully be changed, and Alice's edit will be dropped.
+
+When considering whether a given edit is suitable for a given scenario,
```

Comment:
No. Preconditions determine whether an edit is valid in a given situation or not. This sentence is about the suitability of an edit for a given scenario given what those preconditions are.

Reply:
If somethign is invalid then it will not be allowed to happen. By contrast, if something is not suitable, it will be allowed to happen but there was possbily something better to do.